lib/CoffeeTags/parser.rb

```diff
@@ -130,31 +178,61 @@ def execute!
         [
           [@class_regex, 'c'],
           [@proto_meths, 'p'],
+          [@func_regex, 'f'],
           [@var_regex, 'v'],
           [@block, 'b']
         ].each do |regex, kind|
           mt = item_for_regex line, regex, level, :source => line, :line => line_n, :kind => kind
-          @tree << mt unless mt.nil?
+          unless mt.nil?
+            # TODO: one token should not fit for multiple regex
+            classes.push mt if kind == 'c'
+            next if kind == 'f' # wait for later to determine whether it is a class method
+            @tree << mt
+          end
         end
 
-
         # instance variable or iterator (for/in)?
         token = line.match(@token_regex )
         token ||=  line.match(@iterator_regex)
 
         # we have found something!
         if not token.nil?
-          o = {
-            :name => token[1],
-            :level => level,
-            :parent => '',
-            :source => line,
-            :line => line_n
-          }
+          # should find token through the tree first
+          token_name = token[1]
+          existed_token = @tree.find {|o| o[:name] == token_name}
+          if existed_token
+            o = existed_token
+          else
+            o = {
+              :name => token_name,
+              :level => level,
+              :parent => '',
+              :source => line,
+              :line => line_n
+            }
+          end
 
           # remove edge cases for now
           # - if a line containes a line like:  element.getElement('type=[checkbox]').lol()
-          is_in_string = line =~ /.*['"].*#{token[1]}.*=.*["'].*/
+          is_in_string = false
+          token_match_in_line = line.match token_name
+          unless token_match_in_line.nil?
+            offset = token_match_in_line.offset 0
+            str_before = line.slice 0, offset[0]
+            str_after = line.slice offset[1], line.size
+            for s in [str_before, str_after]
```

Comment:
For is not idiomatic ruby, this would look better like this:


```ruby
[ str_before, str_after].map do |str|
is_in_string =   [ '"', '\''].any? { |q| str.scan(q).size % 2 == 1 }
end

```

(or something like that- I'm not 100% I understand the algorithm here :-))